### PR TITLE
flow init commit

### PR DESCRIPTION
**Pull request by PrzemekWisniewski**

### CODE REVIEW

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+/**
+ * Created by przemek on 19.10.2016.
+ */
+
+@Service
+@Slf4j
+public class BaseScheduledFlow {
+
+    private Client client;
+    private String deviceUUID;
```
Comment (tomasz-rojek) [inline src/main/java/com/base/BaseScheduledFlow.java]: keeping state is not necessary here

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    public BaseScheduledFlow() {
+        log.info("BaseScheduledFlow service launching...");
+        deviceUUID = getProperty("deviceUUID");
+        this.client = new Client(new Configuration.Builder()
+                .accessToken(getProperty("BASECRM_ACCESS_TOKEN"))
+                .build());
+    }
+
+    @Scheduled(fixedDelay = 15000)
+    public void reportCurrentTime() {
```
Comment (tomasz-rojek) [inline src/main/java/com/base/BaseScheduledFlow.java]: I don't understand this method's name

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Created by przemek on 19.10.2016.
+ */
+
+@Service
+@Slf4j
+public class BaseScheduledFlow {
```
Comment (tomasz-rojek) [inline src/main/java/com/base/BaseScheduledFlow.java]: this class is too big, should be splitted to more classes/responsibilities

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        MDC.clear();
+        return true;
+    }
+
+    private boolean isDealWon(Deal deal) {
+
+        Optional<Stage> result = client.stages()
+                .list(new StagesService.SearchCriteria().active(false))
+                .stream()
+                .filter(stage -> deal.getStageId().equals(stage.getId()) && "won".equals(stage.getCategory()))
```
Comment (ppietr) [inline src/main/java/com/base/BaseScheduledFlow.java]: please extract it to method

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        LocalDate now = LocalDate.now();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
+
+        String dealName = contact.getName() + " " + now.format(formatter);
+        newDeal.setName(dealName);
+
+        Deal createdDeal = client.deals().create(newDeal);
+        log.info("created deal with id: {}, name: {}, owned by: {}, assigned to contact: {}", createdDeal.getId(), dealName, contact.getOwnerId(), contact.getId());
+
+        return null != createdDeal;
```
Comment (ppietr) [inline src/main/java/com/base/BaseScheduledFlow.java]: is it possible ?

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Created by przemek on 19.10.2016.
+ */
+
+@Service
```
Comment (volphy) [inline src/main/java/com/base/BaseScheduledFlow.java:19]: This class can be annotated as @Component. However, annotating it as @Service is fine.

File: src/main/java/com/base/BaseScheduledFlow.java (hunk 3 above)
Comment (volphy) [reply src/main/java/com/base/BaseScheduledFlow.java]: Does this class require public visibility?

Default visibility (no scope qualifier) should suffice here.

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        MDC.put("deal", deal.getId().toString());
+        String syncEventType = meta.getSync().getEventType();
+        boolean isDealUpdated = "updated".equals(syncEventType);
+
+        if (isDealUpdated) {
+            log.info("processing deal, sync event received: '{}'", syncEventType);
+            if (isDealWon(deal)) {
+                Contact relatedContact = client.contacts().get(deal.getContactId());
+                log.info("checking contact {} for reassigning", relatedContact.getId());
+                if (isCurrentOwnerASalesRepUser(getCurrentUser(relatedContact.getOwnerId()))) {
```
Comment (ppietr) [inline src/main/java/com/base/BaseScheduledFlow.java]: please extract it

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        } else {
+            log.info("processing contact, received sync event != 'created'. Nothing to do");
+        }
+
+        clearSync(meta);
+        MDC.clear();
+        return true;
+    }
+
+    private void clearSync(Meta meta) {
```
Comment (volphy) [inline src/main/java/com/base/BaseScheduledFlow.java]: Is this method necessary?

AFAIK, if using Java API acknowledgments are implicitly sent anyway.

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+                .ack(deviceUUID, Arrays.asList(meta.getSync().getAckKey()));
+    }
+
+    private boolean verifyContact(final Contact contact) {
+        boolean isContactAnOrganization = contact.getIsOrganization();
+        log.info("verifying contact. contact is a company: {}", isContactAnOrganization);
+
+        return verifyNoExistingDealsOnContact(contact)
+                && isCurrentOwnerASalesRepUser(getCurrentUser(contact.getOwnerId()))
+                && contact.getIsOrganization();
```
Comment (ppietr) [inline src/main/java/com/base/BaseScheduledFlow.java]: isContactAnOrganization ?

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+    }
+
+    private User getCurrentUser(Long id) {
+        return client.users().get(id);
+    }
+
+    private boolean verifyNoExistingDealsOnContact(final Contact contact) {
+        List<Deal> deals = client.deals()
+                .list(new DealsService.SearchCriteria().contactId(contact.getId()));
```
Comment (ppietr) [inline src/main/java/com/base/BaseScheduledFlow.java]: please extract to getDealById method

**Comment by PrzemekWisniewski** [comment]
@villeu-solidbrain @volphy @ppietr 
Done some refactoring. Could you please have a look?

File: src/main/java/com/base/BaseScheduledFlow.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Created by przemek on 19.10.2016.
+ */
+
+@Service
+@Slf4j
+class BaseScheduledFlow {
+
+    private final Client client;
```
Comment (tomasz-rojek) [inline src/main/java/com/base/BaseScheduledFlow.java]: it doesn't need to be a field, let it be constructed each time with @Scheduled

File: src/main/java/com/base/util/BaseAppUtil.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
+package com.base.util;
+
+import com.getbase.Client;
+import com.getbase.Configuration;
+import org.springframework.core.env.MissingRequiredPropertiesException;
+
+/**
+ * Created by przemek on 19.10.2016.
+ */
+public interface BaseAppUtil {
```
Comment (tomasz-rojek) [inline src/main/java/com/base/util/BaseAppUtil.java]: don't use interfaces like that -> make it final concrete class, or use lombok's @UtilityClass

File: src/main/java/com/base/processors/ContactProcessor.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+    private void createDeal(final Contact contact) {
+        final Deal newDeal = new Deal();
+        newDeal.setOwnerId(contact.getOwnerId());
+        newDeal.setContactId(contact.getId());
+
+        LocalDate now = LocalDate.now();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
+
+        String dealName = contact.getName() + " " + now.format(formatter);
```
Comment (ppietr) [inline src/main/java/com/base/processors/ContactProcessor.java]: String getDealName(String name) ?

File: src/main/java/com/base/processors/DealProcessor.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+    private Predicate<Stage> getStagePredicate(Deal deal) {
+        return stage -> deal.getStageId().equals(stage.getId()) && "won".equals(stage.getCategory());
+    }
+
+    private boolean isCurrentOwnerASalesRepUser(Long ownerId) {
+        User currentUser = getCurrentUser(ownerId);
+        String usersStatus = currentUser.getStatus();
+        boolean isUserSalesRep = currentUser.getEmail().contains("_salesrep@");
+        log.info("isUserASalesRepUser, userId: {}, user's status: {}, and user is a sales rep: {}", currentUser.getId(), usersStatus, isUserSalesRep);
```
Comment (ppietr) [inline src/main/java/com/base/processors/DealProcessor.java]: we need to log it ? isUserASalesRepUser

**Pull request merged**